infra/pySourceCredChart/js/lib/pySourceCredChart.js

```diff
@@ -0,0 +1,320 @@
+var widgets = require('@jupyter-widgets/base');
+require('../style/pySourceCredChart.css');
+
+import * as d3 from 'd3';
+
+/// set constants
+const MAX_SIZE_PIXELS = 200
+
+var sourceCredChartModel = widgets.DOMWidgetModel.extend({
+	defaults:_.extend(_.result(this, 'widgets.DOMWidgetModel.prototype.defaults'),{
+		_model_name: 'sourceCredChartModel',
+		_view_name: 'sourceCredChartView',
+		_model_module: 'pySourceCredChart',
+		_view_module: 'pySourceCredChart',
+	})
+});
+
+var sourceCredChartView = widgets.DOMWidgetView.extend({
+	
+	render: function(){
+		var that = this;
+		/// set up chart space without data
+		this.chartContainer = d3.select(this.el);
+		
+		// set up scaffolding
+		d3setScaffolding(that);
+		
+		var message = new Promise(function(resolve, reject){
+			resolve(that.model.get('_model_msg'))
+		});
+		
+		message
+			.then(this.message_changed())
+			.catch(function(err){
+				console.log(err); 
+				});
+		
+		this.model.on('change:_model_msg', this.message_changed, this);
+	},
+	
+	message_changed: function(){
+		this.el.msg = this.model.get('_model_msg');
+		this.draw(this.el.msg[0]);
+	},
+	
+	draw: function(opts){
+		//preserve context for closures
+		var that = this;
+		
+		// parse opts object sent as message
+		this.data = opts.data;
+		this.scores = opts.scores;
+		this.mapping = opts.mapping;
+		this.options = opts.options;
+		
+		//initialize simulation
+		///define variables from data for simulation only
+		///data updates chart and restarts simulation later
+		this.links = this.data[this.mapping.edges];
+		if(that.mapping.source){
+			this.links.forEach(function(d){
+				d.source = d[that.mapping.source];
+				d.target = d[that.mapping.target];
+			});
+		}
+    
+		this.nodes = this.data[this.mapping.nodes];
+		
+		/// define simulation object
+		this.simulation = _simulation(that, this.nodes, this.links);
+    
+		// update chart
+		// this.addZoom();
+		d3updateNodes(that, this.nodes);
+		d3updateEdges(that, this.links);		
+		
+				
+		// Update and restart the simulation.
+		that.simulation.nodes(this.nodes);
+		that.simulation.force("link").links(this.links);
+		that.simulation.alpha(1).restart();
+		
+		function timeIt(){
+		  setTimeout(function(){
+			console.log("call stop");
+			that.simulation.alpha(0).restart();
+			that.simulation.stop();
```

Comment:
why do we set the alpha to 0, then restart it, and then stop it? wouldn't just stopping it be sufficient?

Reply:
also, if you want the simulation to stop on its own over time, there's no need to do this with a timer. you can just set a target alpha lower than the minimum alpha, which is indeed the default settings for a force layout. the alpha values that I chose in the hackathon prototype were specifically chosen because I wanted the simulation to keep running forever.

cf [the docs on alpha decay](https://github.com/d3/d3-force#simulation_alphaDecay)